Add an implementation of Bellman–Ford algorithm

RodionovMaxim05 opened the pull request:
Добавлена реализация алгоритма поиска кратчайшего пути в взвешенном графе:

- [Алгоритм Беллмана — Форда](https://ru.wikipedia.org/wiki/Алгоритм_Беллмана_—_Форда)

RodionovMaxim05 committed:
feat: implement Bellman–Ford algorithm

RodionovMaxim05 committed:
docs: add doc-strings for BellmanFordShortestPath

RodionovMaxim05 committed:
test: add tests for BellmanFordShortestPath

IlyaSuponev committed:
Change style of PR 

* feat: add reporting coverage and tests results on PR review

* refactor: change place 'detekt' gradle plugin implementation

* refactor: update code style by detekt rules

* feat: change main CI github action by split it on two jobs

RodionovMaxim05 committed:
refactor: change BellmanFordShortestPath and TestBellmanFordShortestPath according to detekt`s requirements
Put tabs instead of spaces, specify the type of exceptions, replace if and throw with a require.

RodionovMaxim05 committed:
refactor: BellmanFordShortestPath and TestBellmanFordShortestPath
In the "getPath" function change the return value to a list, move the mapping an index to each vertex into a separate "vertexIndexing" function, add detection of negative cycles consisting of one vertex, edit the documentation and comments.

RodionovMaxim05 committed:
Merge branch 'develop' into dev/BellmanFord

IlyaSuponev commented on graphs-lab/src/main/kotlin/graphs_lab/algs/BellmanFordShortestPath.kt:
```diff
@@ -0,0 +1,152 @@
+		if (matrix.isEmpty()) {
+			vertexIndexing()
+			getPaths(idSource, indexedVertices)
+		} else if (matrix[indexedVertices[idSource]!!][0] == Double.POSITIVE_INFINITY) {
```
Возможно имеет смысл место `!!` использовать `.getOrDefault()`:
```suggestion
		} else if (matrix[indexedVertices.getOrDefault(idSource, -1)][0] == Double.POSITIVE_INFINITY) {
```

`-1` можно заменить на другой неверный индекс.

RodionovMaxim05 replied:
А зачем? Здесь это только уменьшит понимание происходящего.

IlyaSuponev replied:
Ок, но в большей степени я просто предложил избавиться от `!!`.

IlyaSuponev commented on graphs-lab/src/main/kotlin/graphs_lab/algs/BellmanFordShortestPath.kt:
```diff
@@ -0,0 +1,152 @@
+		// Finding the length of the shortest path and the number of edges in this path
+		weightOfShortestPath = Double.POSITIVE_INFINITY
+		var numberOfEdgesInThePath = 0
+		var tempIdTarget = indexedVertices[idTarget]!!
```
Та же проблема:
```suggestion
		var tempIdTarget = indexedVertices.getOrDefault(idTarget, -1)
```
`-1` можно заменить на другой неверный индекс.
Есть ещё такие случаи, но они полностью аналогичны, поэтому я их далее не отмечу.

IlyaSuponev commented on graphs-lab/src/test/kotlin/graphs_lab/algs/TestBellmanFordShortestPath.kt:
Тесты хорошие, но, по моему мнению, надо накатить тесты для неориентированного графа.

RodionovMaxim05 replied:
Накатил)

RodionovMaxim05 committed:
refactor: BellmanFordShortestPath and TestBellmanFordShortestPath
The return value of the "getPath" function is changed to “null”, if there is no path between the vertices,  and because of this, the return value of the entire function is changed to “List<I>?”. Remove unnecessary inheritance "I" from "Any".

RodionovMaxim05 committed:
test: add new tests BellmanFordShortestPath
Add tests for undirected graph.

IlyaSuponev commented:
Всё отлично. Хорошая работа.
